src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1134,204 +1124,320 @@ def process_declarations(self, parent, nodes, arg_list):
+                sym.interface = interface
+
+    def process_declarations(self, parent, nodes, arg_list):
+        '''
+        Transform the variable declarations in the fparser2 parse tree into
+        symbols in the symbol table of the PSyIR parent node.
+
+        :param parent: PSyIR node in which to insert the symbols found.
+        :type parent: :py:class:`psyclone.psyGen.KernelSchedule`
+        :param nodes: fparser2 AST nodes to search for declaration statements.
+        :type nodes: list of :py:class:`fparser.two.utils.Base`
+        :param arg_list: fparser2 AST node containing the argument list.
+        :type arg_list: :py:class:`fparser.Fortran2003.Dummy_Arg_List`
+
+        :raises NotImplementedError: the provided declarations contain \
+                                     attributes which are not supported yet.
+        :raises GenerationError: if the parse tree for a USE statement does \
+                                 not have the expected structure.
+        :raises SymbolError: if a declaration is found for a Symbol that is \
+                    already in the symbol table with a defined interface.
+        :raises InternalError: if the provided declaration is an unexpected \
+                               or invalid fparser or Fortran expression.
+
+        '''
+        # Search for any accessibility statements (e.g. "PUBLIC :: my_var") to
+        # determine the default accessibility of symbols as well as identifying
+        # those that are explicitly declared as public or private.
+        (default_visibility, explicit_public_symbols,
+         explicit_private_symbols) = self._parse_access_statements(nodes)
+
+        # Look at any USE statements
+        self._process_use_stmts(parent, nodes)
+
+        for decl in walk(nodes, Fortran2003.Type_Declaration_Stmt):
+            try:
+                self._process_decln(parent, decl,
+                                    default_visibility,
+                                    explicit_public_symbols,
+                                    explicit_private_symbols)
+            except NotImplementedError:
+                # Found an unsupported variable declaration. Create a
+                # DataSymbol with UnknownType for each entity being declared.
+                orig_children = list(decl.children[2].children[:])
+                for child in orig_children:
+                    # Modify the fparser2 parse tree so that it only declares
+                    # the current entity
+                    decl.children[2].items = (child,)
+                    symbol_name = str(child.children[0])
+                    # If a declaration declares multiple entities, it's
+                    # possible that some may have already been processed
+                    # successfully and thus be in the symbol table.
+                    try:
+                        parent.symbol_table.add(
+                            DataSymbol(symbol_name, UnknownType(str(decl))))
```

Comment:
I think I wanted to move away from depending on the existence of the parse tree.

Reply:
OK. In the longer term I am thinking that a CodeBlock might have an interface associated with it that help us determine some things that go on inside it e.g. what variables are accessed and how they are accessed. It may be that we might want/need the same sort of thing for an UnknownType but I can't think of what it would be. Anyway it would be easy to change back if needed and we have no use case so I'm happy with the implementation as it is.